### ISSUE TITLE
Indent rules in editor fight me

### ISSUE THREAD

**Issue by abarth** [flutter/flutter-intellij#675]
## Steps to Reproduce

1. git clone git@github.com:flutter/flutter.git
2. idea flutter
3. Open container.dart
4. Put your cursor on the end of line 182 (where the XXXX is below:

```dart
    if (foregroundDecoration != null) {
      current = new DecoratedBox(
        decoration: foregroundDecoration,
        position: DecorationPosition.foreground,XXXXX
```

5. Press the `<enter>` key.
6. The cursor is now in column 10 but I'd like it to be in column 8 (matching the intent on the previous line).
7. Press `<backspace>`
8. The entire line is gone but I'd like either one or two spaces to be deleted.

## Workaround

The workaround I've need to employ is the following key combination whenever I need to make a new line:

`<enter>`, `<shift>-<left-arrow>`, `<shift>-<left-arrow>`, `<delete>`

It's quite painful.

## Version info

```
[✓] IntelliJ IDEA Community Edition (version 2016.3)
    • Dart plugin version 163.9166.22
    • Flutter plugin version 0.1.8.1
```

**Comment by devoncarew** [comment]
@stevemessick, can you comment on this? Are aspects of this addressed by your text editing changes?

**Comment by stevemessick** [comment]
I have not tested this specific example, but this pattern is one I've been working on. It is fixed in the latest EAP version of the Dart plugin source code. I think the code can be cherry-picked into the 2016.3 release so it should be available soon-ish. If you just want to disable smart-backspace you can change it in preferences (Editor>General>Smart Keys). I use "To nearest indent position".

@alexander-doroshko Any update on cherry-picking the patch that makes formatting sensitive to indent-mode?

**Comment by alexander-doroshko** [comment]
@stevemessick I've cherry-picked your fix and published the Dart plugin with the fix for IntelliJ IDEA 2016.3.x. You should get a suggestion to update the plugin or try Help | Check for updates.
